### ISSUE TITLE
Design: state trie pruning

### ISSUE THREAD

**Issue by sorpaas** [ethereumproject/go-ethereum#440]
Currently go-ethereum still doesn't have pruning, which creates some troubles for users with small disks. Those people, if wishing to use go-ethereum, would occasionally need to delete the state trie database and resync with `--fast` again. The later also skips several consensus validation steps and it can be a problem. Below is a draft design that uses a Rust FFI through `etcommon-trie` to implement pruning for go-ethereum.

#### `etcommon-trie`

The current `etcommon-trie` design would need some change, most notably:

* It needs to only take an immutable reference to the underlying database, and it never directly changes the database.
* It emits adding/deleting records and pass it back to the client. It is up to the client's decision whether or not inserting it to a database. For example, if it is in a `call` RPC operation, then the records are discarded. In contrast, if it is in block validation or chain reorg, the records would need to be handled.
* We have four basic operations `get`, `insert`, `delete` and `traverse`.

#### Integration

The FFI works by creating an instance that takes a callback `get` from go-ethereum's leveldb. Trie's `get` and `traverse` are immutable and works as normal. For `insert` and `delete`, we store only 5000 (or less) blocks for handling of chain reorg, and garbage collect all other state tries. It works by:

* Create a separate database "side database" that stores adding/deleting records returned from `etcommon-trie` for the last 5000 blocks. The return value of `etcommon-trie`'s `insert` or `delete` is a tuple with the first item being the root hash, the second item being a list of adding records (a tuple of key and value) and the third item being a list of deleting records. The root hash is stored in the block header, and we store the adding and deleting records to the side database.
* Moving forward of the blockchain, we apply the side database's last item's deleting records (and remove the last item), and apply the first item's adding records.
* Moving backward of the blockchain, we apply the side database's first item's adding records as deleting records (so in this case, it is possible to have the side database with less than 5000 items).

#### Discussion

Besides getting pruning implemented, there're several important properties for this design:

* We don't require any existing database change, so no migration is needed. If users are not satisfied, they can also switch back and continue to use `ethdb`.
* The interface is extremely simple, and we keep algorithms and complexity separated properly.
* This also accidentally fits in SputnikVM's `AccountChange` model, where changes are not required to get immediately written into database, but can optionally be collected and propagated. So with some further optimization, we are also likely to get some performance improvements.
* We introduce no additional toolchain, as we reuse the toolchain already there by `sputnikvm-ffi`. Rust and Go are both doing what they're good at.

**Comment by whilei** [comment]
I am really impressed by the "lightweight"ness of the approach -- requiring no existing database migration, simple interface, and building on the functionality of what already exists.

If we were to implement pruning as an FFI feature, what implications would it have for the possibility of eventually implementing it as a "native Go" feature? Does it preclude or limit us there?

I'm not intuitively sure if adding another "side database" is architecturally a pro or a con -- I like that it seems modular, but dislike that it seems to add a kind of complexity to data location.

**Comment by sorpaas** [comment]
For everyone else's reference, here're the discussion on Slack that answers @whilei's questions:

```
ia 12:05 AM
I would rather not wait and cherry-pick
12:07
And getting pruning going on our geth would be excellent, and the sooner the better
12:07
of course
12:10
And maybe that means now going ahead with your idea and solve for it via FFI, and then be able to (re)implement it in Go using the FFI funtionality as a reference/benchmark
12:10
Do you see any challenges that doing it via Rust+FFI would create for eventually moving the feature to Go?
sorpaas 12:12 AM
No. Through FFI it will be implemented as an optional flag as well.
ia 12:13 AM
Would we be able to (re)use the same database?
sorpaas 12:13 AM
Yep.
12:13
(We didn't change any database schema.) (edited)
sorpaas 12:14 AM
So that's why we use "side database". It keeps compatibility and allows us to move at directions at will.
ia 12:14 AM
Yea, got it
ia 12:17 AM
And just double checking sorpaas - I meant would we be able to reuse same "side database"?
sorpaas 12:18 AM
Should be. I don't see why that cannot be reimplemented in go.
ia 12:18 AM
I think the "side database" is what keeps the integration separate from geth's main state db
12:18
so there's no changes to existing schema, no migration, etc
12:18
just "pluggable"
12:18
correct me if i'm wrong
sorpaas 12:18 AM
You can read the design if you're not sure.
ia 12:19 AM
Did, just making sure I understand correctly
```

**Comment by whilei** [comment]
And more copy+paste with following discussion:

```
tzdybal [11:22] 
I think that using rust implementation as reference is a good idea. Logically this functionality 'belongs' to geth so I totally understand targeting at golang implementation. Actually I have very 'relaxed' (not conservative) approach in regards to binding on abi/ffi/binary level - If it works ok, why not use it? We're using cgo anyways, and in the end it's all about calling functions using c call convention. IMHO it's the natural way. (I'll comment about pruning in few hours.)



tzdybal [17:32] 
If I understand correctly, Go code will invoke Rust code invoking Go code (to access database) - am I right?



----- Today December 22nd, 2017 -----
sorpaas [04:49] 
yes


tzdybal [05:55] 
This is a good reason to implement pruning in go (IMHO) ;) but having reference implementation in Rust is a plus.


tzdybal [06:01] 
@sorpaas you need this functionality anyways in Rust, right? Or where going to use etcommon-trie with geth (because of SputnikVM?)?


tzdybal [06:10] 
And my comment regarding performance - I'm pretty sure that there's lot to do. Importing shows clearly that even when we're not waiting on network peers, CPUs and memory are not fully utilized. But for now I think it's not a big problem (it's not visible during normal work). Btw - @ia what's your load in R-PI3? does it look like geth is to slow, can't keep in sync?


sorpaas [07:07] 
Yep it's being used in etclient and I'm planning to add pruning for it soon.


[07:10] 
Anyway, to be honest, pruning for go-ethereum is really none of my business, and just let me know when you reached a decision for a green light or not. Personally I feel this is a bottleneck and it makes it impossible for me to run a full geth client on any VPS right now (even on `sputnikvm1`), but if Rust is not welcomed in go-ethereum and we feel that this is something that we can wait for someone to rewrite everything again in Go, I would be more than happy and would rather spend more time getting this done for etclient instead.


tzdybal [07:53] 
I think, that pruning is very important... and having this before parity is a great advantage, especially for constrained environments (like cheap vps)


[07:54] 
So I think that even experimental, rust version should be released


bunda [10:37] 
Kinda late, but here is  mine copule cents:
* +1 for go implementation. Can't see any reason to manage zoo of languages in one  projects
* Rust is not so fast as most people saying about it. Especially crypto stuff, they unmature and untested. 
Often lots of people facing creating binding`s into old-good C.
* "... Go code will invoke Rust code invoking Go code (to access database)..." - I would omit such things in production at any cost
Less moving parts - more robust.



bunda [10:44] 
and also seems Go cross-complilation won't work with embedded Rust
(FFI to C definetly brokes it, but haven't investigate about Rust)
```

**Comment by AlekSi** [comment]
I think state pruning was implemented in 1.8?
